### ISSUE TITLE
HTML to honor nearest-neighbor

### ISSUE THREAD

**Issue by tresf** [qzind/tray#57]
HTML rendering currently blurs all images as they are scaled up.  This seems to be an effect of the default scaling that JavaFX `WebView` uses.

A current work-around is to use a library such as `ImageMagic` or `gd` to scale the image up prior to embedding it into the HTML document.

Here's an example of `sample.html`'s printHTML() code, modified to use an up-scaled version of the `assets/img/image_sample.png`, and then hard-coding the dimensions.

![screen shot 2016-05-10 at 10 03 31 pm](https://cloud.githubusercontent.com/assets/6345473/15168012/78cadb0a-16fb-11e6-81e3-57f1dcea1ab0.png)

**Comment by tresf** [comment]
A stop-gap for better HTML barcodes is a library such as `JsBarcode`.

https://github.com/lindell/JsBarcode

![image](https://cloud.githubusercontent.com/assets/6345473/15256669/99b03bf0-1910-11e6-9ff4-daf3b67d042d.png)



![image](https://cloud.githubusercontent.com/assets/6345473/15256644/73758e22-1910-11e6-8306-d93e93d08180.png)

**Comment by vinhlh** [comment]
same problem with jquery barcode

**Comment by vinhlh** [comment]
+1

**Comment by vinhlh** [comment]
do you have any estimation of how long this ticket will be resolved @tresf?

**Comment by tresf** [comment]
@vinhlh this is currently being tested on the `2.1` branch with success, but requires Java 1.8 as we've redesigned the way we do HTML printing in our `2.1` branch.

Compilation/packaging is dead-easy on Linux and Mac (Windows requires a few extra steps): https://qz.io/wiki/compiling

If you have some time to test the `2.1` branch, that will help us towards a RC release.

If you do not have an environment suitable for building, ping us at support@qz.io and we'll fire over an installer.

**Comment by vinhlh** [comment]
Yep, I had the same problem with you on Mac OS environment on `2.1` branch build.

**Comment by tresf** [comment]
I had forgotten to mention, you'll have to turn rasterize off.

**Comment by vinhlh** [comment]
@tresf 

I tried with rasterize = false, but the result is still the same
I added some debug code to validate the image's quality. Tt's good, but on paper, texts and barcodes are bad
![](http://i.imgur.com/mi2ZehO.jpg)

this is direct print version
![](http://i.imgur.com/AFO2aX5.jpg)
```java
    public void print(PrintOutput output, PrintOptions options) throws PrinterException {
        if (options.getPixelOptions().isRasterize()) {
            //grab a snapshot of the pages for PrintImage instead of printing directly
            for(WebAppModel model : models) {
                try {
                    File outputfile = new File("/Users/vinhlh/Desktop/raster.png");
                    ImageIO.write(WebApp.raster(model), "png", outputfile);
                    // System.exit(1);
                }
                catch(Throwable throwable) {
                    throwable.printStackTrace();
                }
```

**Comment by tresf** [comment]
@vinhlh, albeit mostly unrelated to this bug report I'm fairly certain the Mac print libraries are incapable of proper thermal printing due to an upstream bug.

Here's where we tested against this bug quite extensively: https://github.com/qzind/tray/issues/38#issuecomment-215884556

Our workaround for Mac is done here https://github.com/qzind/tray/commit/7b8d27c7eae7a1a7315065f7bd4b8b4b644c9216 but only applies to images.  JavaFX [doesn't have a `nearest-neighbor` setting, so quality is always lost when using HTML,  unfortunately](http://stackoverflow.com/questions/37797351/).

2.1 fixes this quite well by using JavaFX's internal print libraries, but due to the Mac bug, I think it's still broken.

https://community.oracle.com/thread/3912511

Anyway, it's on our radar to report directly to Oracle, and we will once we get a benchmark on the time it takes for them to fix the PDFBOX hard-crash, which we have an open Service Request for.

**Comment by tresf** [comment]
Ok, update.... If I use jquery_barcode, the barcode looks really good.  Attached is 2.1 printing to a Zebra LP2844 on OS X 10.11 El Capitan.  Sample code here: [jquery_barcode.zip](https://github.com/qzind/tray/files/414610/jquery_barcode.zip)

![screen shot 2016-08-12 at 12 10 38 am](https://cloud.githubusercontent.com/assets/6345473/17612549/41af61a8-6021-11e6-9044-bc2b0023d6e2.png)

**Comment by vinhlh** [comment]
@tresf 
it seems your printing version is quite different with the original version.
![](http://i.imgur.com/ZjhUlTx.png)

**Comment by tresf** [comment]
@vinhlh yes, there is.  This was only meant to illustrate the print quality, not layout accuracy of the JavaFX engine.

**Comment by tresf** [comment]
@vinhlh here's a better example.  Apparently JavaFX was zoomed out too much causing some strange rendering with the divs.  This one does't need the special 4x6 dimensions.

[jquery-barcode2.zip](https://github.com/qzind/tray/files/414648/jquery-barcode2.zip)

![screen shot 2016-08-12 at 1 00 57 am](https://cloud.githubusercontent.com/assets/6345473/17613166/41d23e7e-6028-11e6-9585-7f786908085a.png)

**Comment by tresf** [comment]
Closed via https://github.com/qzind/tray/commit/9f47456a02f4a49b22db71c99c2098c16e5140e9 (2.1 branch)

**Comment by vinhlh** [comment]
@tresf  using scaling WebView (instead of setZoom with 1.0) solves my barcode accuracy's problem.

**Comment by tresf** [comment]
> @tresf using scaling WebView (instead of setZoom != 1.0) solves my barcode accuracy's problem.

2.0 or 2.1?  Can you elaborate a bit more?

**Comment by vinhlh** [comment]
2.1, when using jquery barcode with width = 1px, just make sure that don't use setZoom < 1.0. this will create the barcode accuracy's problem.
instead, can use this function `webView.getTransforms().add(new Scale(0.75, 0.75));` to scale the webview.

**Comment by tresf** [comment]
@vinhlh opening a new issue for this, thanks.

**Comment by tresf** [comment]
@vinhlh FYI, implemented your suggestion on 2.1 branch per https://github.com/qzind/tray/commit/dd1451f4cbfc0d1ae4e037a80ff806a8cdc9baee.
